File: December 2022/Day 8/C++ Code/3 Divisors.cpp
// Problem Link :- https://practice.geeksforgeeks.org/problems/3-divisors3942/1
// Video Solution Link :- https://youtu.be/Nv6NtSk4igU

//{ Driver Code Starts
#include<bits/stdc++.h> 
using namespace std; 

// } Driver Code Ends
class Solution
{
public:
    // Time Complexity :- O(q*N*log(log(N)));
    // Space Complexity :- O(N), where N is min(10^6,N);
    vector<int> threeDivisors(vector<long long> query, int q)
    {
        // Write your code here
        vector<int> prime = primeNumber(1000000);
        vector<int> ans;
        for(long q1 : query) {
            ans.push_back(solve(prime,q1));
        }
        return ans;
    }
    
    vector<int> primeNumber(int n) {
        vector<int> prime(n+1,1);
        prime[0]=0;
        prime[1]=0;
        
        for(int i=2;i<=n;i++) {
            if(prime[i]==1) {
                for(int j=i+i;j<=n;j+=i) {
                    prime[j]=0;
                }
            }
        }
        vector<int> list;
        for(int i=0;i<=n;i++) {
            if(prime[i]==1) list.push_back(i);
        }
        return list;
    }
    
    int solve(vector<int> prime,long long q) {
        int v1 = (int)sqrt(q);
        int lo=0,hi=prime.size()-1;
        int ans=0;
        while(lo<=hi) {
            int mid = lo+(hi-lo)/2;
            if(prime[mid]<=v1) {
                lo=mid+1;
                ans = max(ans,mid+1);
            } else {
                hi=mid-1;
            }
        }
        return ans;
    }
};

//{ Driver Code Starts.
int main()
{
    int t;
    cin >> t;
    while(t--){
        int q;cin>>q;
        vector<long long> query(q);
        for(int i=0;i<q;i++){
            cin>>query[i];
        }
        Solution ob;
            
        vector<int> ans = ob.threeDivisors(query,q);
        for(auto cnt : ans) {
            cout<< cnt << endl;
        }
    }
    return 0;
}
// } Driver Code Ends

File: December 2022/Day 8/JAVA Code/3 Divisors.java
Problem Link :- https://practice.geeksforgeeks.org/problems/3-divisors3942/1
Video Solution Link :- https://www.youtube.com/channel/UC1P9bWDLTPS1Zp03m5PQEoQ

//{ Driver Code Starts
//Initial Template for Java
import java.io.*;
import java.util.*;

class GFG
{
    public static void main(String args[])throws IOException
    {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while(t-->0){
            int q = sc.nextInt();
            ArrayList<Long> query = new ArrayList<>();
            for(int i=0;i<q;i++){
                query.add(sc.nextLong());
            }
            Solution ob = new Solution();
                
            ArrayList<Integer> ans = ob.threeDivisors(query,q);
            for(int cnt : ans) {
                System.out.println(cnt);
            }
        }
    }
}

// } Driver Code Ends


//User function Template for Java
class Solution{
    // Time Complexity :- O(q*N*log(log(N)));
    // Space Complexity :- O(N), where N is min(10^6,N);
    static ArrayList<Integer> threeDivisors(ArrayList<Long> query, int q){
        // code here
        ArrayList<Integer> prime = primeNumber(1000000);
        ArrayList<Integer> ans = new ArrayList<>();
        for(long q1 : query) {
            ans.add(solve(prime,q1));
        }
        return ans;
    }
    
    public static ArrayList<Integer> primeNumber(int n) {
        int[] prime = new int[n+1];
        Arrays.fill(prime,1);
        prime[0]=0;
        prime[1]=0;
        
        for(int i=2;i<=n;i++) {
            if(prime[i]==1) {
                for(int j=i+i;j<=n;j+=i) {
                    prime[j]=0;
                }
            }
        }
        ArrayList<Integer> list = new ArrayList<>();
        for(int i=0;i<=n;i++) {
            if(prime[i]==1) list.add(i);
        }
        return list;
    }
    
    public static int solve(ArrayList<Integer> prime,long q) {
        int v1 = (int)Math.sqrt(q);
        int lo=0,hi=prime.size()-1;
        int ans=0;
        while(lo<=hi) {
            int mid = lo+(hi-lo)/2;
            if(prime.get(mid)<=v1) {
                lo=mid+1;
                ans = Math.max(ans,mid+1);
            } else {
                hi=mid-1;
            }
        }
        return ans;
    }
}
